src/lib/Component/Renderer/DefaultRenderer.php

```diff
@@ -57,15 +57,28 @@ public function renderGroup(string $groupName, array $parameters = []): array
      */
     public function renderSingle(string $groupName, string $name, array $parameters = []): string
     {
+        $components = $this->registry->getComponents($groupName);
+
+        return $this->renderSingleWithComponents($groupName, $name, $components, $parameters);
+    }
+
+    /**
+     * @param array<mixed> $parameters
+     * @param array<string, \Ibexa\Contracts\TwigComponents\ComponentInterface> $components
+     */
+    private function renderSingleWithComponents(
+        string $groupName,
+        string $name,
+        array $components,
+        array $parameters = []
+    ): string {
         $this->eventDispatcher->dispatch(new RenderSingleEvent(
             $this->registry,
             $groupName,
             $name,
             $parameters
         ), RenderSingleEvent::class);
 
-        $components = $this->registry->getComponents($groupName);
-
```

Comment:
You basically moved acquisition of components before the event. I don't think it's worthy of a dedicated method, which is only used in this particular instance?

Also, why?